mp/src/game/shared/momentum/run/mom_run_safeguards.cpp

```diff
@@ -0,0 +1,180 @@
+    m_arrpszAction[RUN_SAFEGUARD_SAVELOC_TELE] = "teleport to saveloc";
+    m_arrpszAction[RUN_SAFEGUARD_CHAT_OPEN] = "open chat";
+
+    for (int i = 0; i < RUN_SAFEGUARD_COUNT; i++)
+    {
+        m_szLastTimesPressed[i] = 0.0f;
+        m_szDoublePressSafeguards[i] = true;
+    }
+}
+
+bool MomRunSafeguards::IsSafeguarded(RunSafeguardType_t type)
+{
+    return IsSafeguarded(type, GetModeFromType(type));
+}
+
+bool MomRunSafeguards::IsSafeguarded(RunSafeguardType_t type, RunSafeguardMode_t mode)
+{
+    CMomentumPlayer *pPlayer;
+#ifdef CLIENT_DLL
+    pPlayer = C_MomentumPlayer::GetLocalMomPlayer();
+#elif GAME_DLL
+    pPlayer = CMomentumPlayer::GetLocalPlayer();
+#endif
+    if (!pPlayer)
+        return false; 
+
+    CMomRunEntityData *pEntData;
+#ifdef CLIENT_DLL
+    pEntData = pPlayer->GetCurrentUIEntData();
+#elif GAME_DLL
+    pEntData = pPlayer->GetRunEntData();
+#endif
+    if (!pEntData)
+        return false;
+
+    if (!pEntData->m_bTimerRunning || pPlayer->IsObserver())
+        return false; 
+
+    m_pLocalPlayer = pPlayer;
+
+    switch (mode)
+    {
+    case RUN_SAFEGUARD_MODE_MOVEMENTKEYS:
+        return IsMovementKeysSafeguarded(type);
+    case RUN_SAFEGUARD_MODE_DOUBLEPRESS:
+        return IsDoublePressSafeguarded(type);
+    default:
+        return false;
+    }
+}
+
+RunSafeguardMode_t MomRunSafeguards::GetModeFromType(int type)
+{
+    return GetModeFromType(RunSafeguardType_t(type));
+}
+
+RunSafeguardMode_t MomRunSafeguards::GetModeFromType(RunSafeguardType_t type)
+{
+    switch (type)
+    {
+    case RUN_SAFEGUARD_PRACTICEMODE:
+        return RunSafeguardMode_t(mom_run_safeguard_practicemode.GetInt());
+    case RUN_SAFEGUARD_RESTART:
+        return RunSafeguardMode_t(mom_run_safeguard_restart.GetInt());
+    case RUN_SAFEGUARD_SAVELOC_TELE:
+        return RunSafeguardMode_t(mom_run_safeguard_saveloc_tele.GetInt());
+    case RUN_SAFEGUARD_CHAT_OPEN:
+        return RunSafeguardMode_t(mom_run_safeguard_chat_open.GetInt());
+    default:
+        return RUN_SAFEGUARD_MODE_INVALID;
+    }
+}
+
+void MomRunSafeguards::SetMode(int type, int mode)
+{
+    SetMode(RunSafeguardType_t(type), RunSafeguardMode_t(mode));
+}
+
+void MomRunSafeguards::SetMode(RunSafeguardType_t type, RunSafeguardMode_t mode)
+{
+    switch (type)
+    {
+    case RUN_SAFEGUARD_PRACTICEMODE:
+        mom_run_safeguard_practicemode.SetValue(mode);
+        break;
+    case RUN_SAFEGUARD_RESTART:
+        mom_run_safeguard_restart.SetValue(mode);
+        break;
+    case RUN_SAFEGUARD_SAVELOC_TELE:
+        mom_run_safeguard_saveloc_tele.SetValue(mode);
+        break;
+    case RUN_SAFEGUARD_CHAT_OPEN:
+        mom_run_safeguard_chat_open.SetValue(mode);
+        break;
+    }
+}
+
+bool MomRunSafeguards::IsMovementKeysSafeguarded(RunSafeguardType_t type)
+{
+    int nButtons = 0;
+#ifdef CLIENT_DLL
+    nButtons = ::input->GetButtonBits(engine->IsPlayingDemo());
+#elif GAME_DLL
+    nButtons = m_pLocalPlayer->m_nButtons;
```

Comment:
`nButtons` should be passed into this method from the case statement, getting rid of needing to store the local player in this system.